### ISSUE TITLE
FR: Delete to Word End (ignoring single white space characters)

### ISSUE THREAD

**Issue by fgimian** [vsch/MissingInActions#18]
Hey there, really cool extension, thanks a lot for writing it! 😄 

I have noticed that PyCharm's default "Delete to Word End" action treats a single whitespace (`" "` or `"\t"`) as a word which is not the way other applications work on macOS.

e.g. Suppose we have the text `Hello there my friend.` and suppose we place the cursor at the beginning (just before the letter H) and start hitting Option + Delete, this is what we see:

**TextEdit / Sublime Text**

```
Hello there my friend.
 there my friend.
 my friend.
 friend.
.
```

In PyCharm, we see:

```
Hello there my friend.
 there my friend.
there my friend.
 my friend.
my friend.
 friend.
friend.
.
```

In macOS apps, a single space or tab between words is not treated as a word, only multiple spaces are.

I was wondering if you would consider writing a new action "Delete to Word End Excluding Single Whitespace" so that we can map Option + Delete to this instead of the default PyCharm behaviour.

Thanks so much for your time!
Fotis

**Comment by vsch** [comment]
@fgimian, I will add the action for compatibility with existing app behavior.

I find all of these behaviors, sublime text and the IDE, weird and requiring way too many keystrokes to quickly delete a chunk of unnecessary code.

I would think the following scenarios would make more sense, I added extra spaces between words to emphasize how multiple spaces are addressed.

`|` shows caret position, with the next line being the result of using the **Delete To Word End** on the previous line\:

Beginning of line, which is treated as caret preceded by whitespace:

```
|Hello   there  my  friend.
|there  my  friend.
|my  friend.
|friend.
|
```

Caret is preceded by whitespace:

```
Hello |  there  my  friend.
Hello |my  friend.
Hello |friend.
Hello |
```

However if the caret is preceded by non-whitespace then behavior is different:

```
He|llo   there  my  friend.
He| there  my  friend.
He| my  friend.
He| friend.
He|
```

Which I think is a more intuitive way to delete to end of word without un-intentionally splicing words together while at the same time not wasting keystrokes to delete intervening whitespace.

**Comment by vsch** [comment]
@fgimian, for symmetry backspace should have a corresponding action: **Backspace to Word Start**

`|` shows caret position, with the next line being the result of using the **Backspace To Word Start** on the previous line\:

End of line, considered as caret followed by whitespace:

```
Hello   there  my  friend. |
Hello   there  my|
Hello   there|
Hello|
|
```

Caret followed by whitespace:

```
Hello   there  my | friend.
Hello   there| friend.
Hello| friend.
| friend.
```

Caret not followed by whitespace:

```
Hello   there  my  fri|end.
Hello   there  my |end.
Hello   there |end.
Hello |end.
|end.
```

```
Hello   there  my  |friend.
Hello   there |friend.
Hello |friend.
|friend.
```

**Comment by vsch** [comment]
@fgimian,

* Add: customizable backspace and delete actions with variations: spaces, alternating word, word
  excluding 1 space, and word. All are customizable with regex for selecting what is deleted.

Available if you download the [distribution zip from this repo](https://github.com/vsch/MissingInActions/raw/master/dist/MissingInActions.1.4.9.zip) and install plugin from disk. I will not be making the official release for a while because I want to polish up the `Bulk Search/Replace` functionality. I added minimal functionality needed for a project undergoing heavy refactoring and doing bulk replace was easier than hunting for each value separately. However, it could use a lot more spit and polish.

The action you want is `Custom Delete Word Excluding Single Space` the default definition for it is as you described. 

<img width="957" alt="screen shot 2018-04-29 at 7 13 39 pm" src="https://user-images.githubusercontent.com/10299086/39412056-35864d6e-4be4-11e8-864e-ea55cccdba9d.png">

All the customizable backspace/delete actions are configured through RegEx in settings with creature comforts added for validating their behaviour in the UI. When the custom regex field has focus, the table result column reflects delete/backspace behaviour of the regex. It repeats delete/backspace action until nothing is deleted or nothing left to delete, showing result of each application as a separate line. Sort of like your example except with highlight of what will be deleted and automatic.

Backspace uses my ReverseRegEx library, it reverses the regex and applies it to a reversed sequence so the regex will be executed backwards. For simple expression the regex remains the same. For more complex ones some tweaking is needed to account for the reversal.

I added your sample to validate the regex:

<img width="1305" alt="screen shot 2018-04-29 at 7 13 04 pm" src="https://user-images.githubusercontent.com/10299086/39412039-d0de1734-4be3-11e8-9b09-fb89e9146fe0.png">

An orthogonally defined backspace is also defined:

<img width="1455" alt="screen shot 2018-04-29 at 7 14 27 pm" src="https://user-images.githubusercontent.com/10299086/39412057-38d448c2-4be4-11e8-825b-d7f2b5ab8b73.png">

Thanks for the request. It motivated me to add the Delete Word/Backspace Word alternatives that I would prefer at the same time and make all of these customizable so anyone cat tweak the functionality to their liking.
